File: 0493-reverse-pairs/0493-reverse-pairs.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int reversePairs(vector<int>& nums) {
        return mergeSort(nums, 0, nums.size() - 1);
    }

private:
    int mergeSort(vector<int>& arr, int low, int high) {
        if (low >= high) return 0;
        int cnt = 0;
        int mid = (low + high) / 2;
        cnt += mergeSort(arr, low, mid);
        cnt += mergeSort(arr, mid + 1, high);
        cnt += countPairs(arr, low, mid, high);
        merge(arr, low, mid, high);
        return cnt;
    }

    int countPairs(vector<int>& arr, int low, int mid, int high) {
        int cnt = 0;
        int right = mid + 1;
        for (int i = low; i <= mid; ++i) {
            while (right <= high && (long long)arr[i] > 2LL * arr[right]) {
                right++;
            }
            cnt += right - (mid + 1);
        }
        return cnt;
    }

    void merge(vector<int>& arr, int low, int mid, int high) {
        vector<int> temp;
        int left = low;
        int right = mid + 1;

        while (left <= mid && right <= high) {
            if (arr[left] <= arr[right]) {
                temp.push_back(arr[left]);
                left++;
            } else {
                temp.push_back(arr[right]);
                right++;
            }
        }

        while (left <= mid) {
            temp.push_back(arr[left]);
            left++;
        }

        while (right <= high) {
            temp.push_back(arr[right]);
            right++;
        }

        for (int i = low; i <= high; ++i) {
            arr[i] = temp[i - low];
        }
    }
};

File: 0493-reverse-pairs/0493-reverse-pairs.py
class Solution(object):

    def mergesort(self, arr,low, high):
        if low >= high:
            return 0
        cnt = 0
        mid = (low + high) // 2
        cnt += self.mergesort(arr, low, mid)
        cnt += self.mergesort(arr, mid+1, high)
        cnt += self.count_pairs(arr, low, mid, high)
        self.merge(arr,low, mid, high)
        return cnt

    
    def merge(self, arr, low , mid, high):
        temp = []
        left = low
        right = mid+1
        while(left <= mid and right <= high):
            if arr[left] <= arr[right]:
                temp.append(arr[left])
                left += 1
            else:
                temp.append(arr[right])
                right += 1

        while left <= mid:
            temp.append(arr[left])
            left += 1
        while right <= high:
            temp.append(arr[right])
            right += 1

        arr[low:high + 1] = temp


    def count_pairs(self, arr, low, mid, high):
        cnt = 0
        right = mid + 1
        for i in range(low, mid + 1):
            while(right <= high and arr[i] > 2 * arr[right] ):
                right += 1

            cnt += right - (mid + 1)
        return cnt




    def reversePairs(self, nums):
        return self.mergesort(nums, 0, len(nums)-1 )
    
        